java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+@file:Suppress("UNCHECKED_CAST")
+
+package arcs.core.data.expression
+
+/**
+ * Traverses a tree of [Expression] objects, evaluating each node, and returning a final
+ * result. Note that types cannot be enforced statically, so it is possible for both
+ * type cast errors and field or parameter errors to occur during evaluation. These will be
+ * reflected as exceptions.
+ */
+class ExpressionEvaluator(val parameterScope: Expression.Scope = ParameterScope()) :
+    Expression.Visitor<Any> {
```

Comment:
I'm a bit wary of having visitors over `Any` type. I haven't reviewed the whole PR, so I'm not sure what might be a good fix.

Reply:
Just read the KDoc: why can't types be enforced statically?

Reply:
There's not really a good fix since Kotlin doesn't have union types and Boolean and Number don't share a common type hierarchy. 

So the workaround would be to return a data class to hold the result to mimic a union, but you'd just be moving the 'Any' and casting to the result holder. That is, I could make 'data class Result(...)' and then there's be no Any in the visit() return value, but the Result() class itself would now have an Any, so at the access sites that extract the value, you'd still need casts, e.g. Result.toBoolean(), Result.toNumber(), etc.

I went this route and it make the code harder to read, created far more object allocation during evaluation, and didn't increase safety that much. I'd prefer to vet the type safety of the expressions during parsing/construction than during execution.

Reply:
As an example, BinaryExpressions can be Number,Number -> Number, or Number,Number -> Boolean, or Any,Any -> Boolean. 

So if you have expr op expr, you in effect have.  op(visitBinary(left), visitBinary(right)), but the visitBinary() can return Number, Boolean, or Any, depending on the BinaryOp.  Kotlin type system and checker isn't powerful enough to encode and check this.

Reply:
If you introduced 3 kinds of BinaryOps you could remove the casts, at the cost of introducing 3 new visitor methods, and this would increase everytime you introduce a new binary op (T x S) -> R that didn't fall into the existing three.